### PR TITLE
Add ruby 2.3 squiggly heredoc support

### PR DESCRIPTION
**Pull request by jjoos**
Fixing #218

### CODE REVIEW

File: lib/ruby_lexer.rb [hunk 1]
```diff
@@ -22,6 +22,7 @@ class RubyLexer
   STR_FUNC_QWORDS = 0x08
   STR_FUNC_SYMBOL = 0x10
   STR_FUNC_INDENT = 0x20 # <<-HEREDOC
+  STR_FUNC_ICNTNT = 0x40 # <<~HEREDOC
```
Comment (zenspider) [inline lib/ruby_lexer.rb]: For the life of me I can't figure out what "ICNTNT" is supposed to mean?

File: lib/ruby_lexer.rb [hunk 2]
```diff
@@ -146,10 +147,11 @@ def expr_result token, text
...
 
-    indent  = (func & STR_FUNC_INDENT) != 0 ? "[ \t]*" : nil
-    expand  = (func & STR_FUNC_EXPAND) != 0
-    eos_re  = /#{indent}#{Regexp.escape eos}(\r*\n|\z)/
-    err_msg = "can't match #{eos_re.inspect} anywhere in "
+    indent         = (func & STR_FUNC_INDENT) != 0 ? "[ \t]*" : nil
+    content_indent = (func & STR_FUNC_ICNTNT) != 0
+    expand         = (func & STR_FUNC_EXPAND) != 0
+    eos_re         = /#{indent}#{Regexp.escape eos}(\r*\n|\z)/
+    err_msg        = "can't match #{eos_re.inspect} anywhere in "
```
Comment (zenspider) [inline lib/ruby_lexer.rb]: If you're going to add a new item that changes alignment, do it in two commits... one where the reviewer can look at content changes, and the other that can be glossed over as reformatting. Your single commit means I need to mentally refine the diff.

File: lib/ruby_lexer.rb [hunk 3]
```diff
@@ -195,17 +197,28 @@ def heredoc here # TODO: rewrite / remove
...
 
-    return :tSTRING_CONTENT, string_buffer.join.delete("\r")
+    string_content = string_buffer.join.delete("\r")
+
+    if content_indent && ruby23?
+      indent = string_content.scan(/^[ \t]*(?=\S)/).min.size || 0
+      string_content = string_content.gsub(/^[ \t]{#{indent}}/, '')
+    end
+
+    return :tSTRING_CONTENT, string_content
```
Comment (zenspider) [inline lib/ruby_lexer.rb]: cool

File: test/test_ruby_lexer.rb [hunk 4]
```diff
@@ -1140,6 +1140,19 @@ def test_yylex_heredoc_double_dash
...
+    setup_lexer_class Ruby23Parser
+
+    assert_lex3("a = <<~\"EOF\"\n  blah blah\n  EOF\n\n",
+                nil,
+                :tIDENTIFIER,     "a",             :expr_cmdarg,
+                :tEQL,            "=",             :expr_beg,
+                :tSTRING_BEG,     "\"",            :expr_beg,
+                :tSTRING_CONTENT, "blah blah\n", :expr_beg,
+                :tSTRING_END,     "EOF",           :expr_end,
+                :tNL,             nil,             :expr_beg)
```
Comment (zenspider) [inline test/test_ruby_lexer.rb]: bad alignment

File: test/test_ruby_lexer.rb [hunk 5]
```diff
@@ -1260,6 +1286,19 @@ def test_yylex_heredoc_single_dash
...
+    setup_lexer_class Ruby23Parser
+
+    assert_lex3("a = <<~'EOF'\n  blah blah\n  EOF\n\n",
+                nil,
+                :tIDENTIFIER,     "a",             :expr_cmdarg,
+                :tEQL,            "=",             :expr_beg,
+                :tSTRING_BEG,     "\"",            :expr_beg,
+                :tSTRING_CONTENT, "blah blah\n", :expr_beg,
+                :tSTRING_END,     "EOF",           :expr_end,
+                :tNL,             nil,             :expr_beg)
```
Comment (zenspider) [inline test/test_ruby_lexer.rb]: ditto

**Comment by zenspider** [comment]
Looking at the implementation in MRI, I might have to reject or to only accept the tests... it's a LOT more complicated :( For example, look at `parser_update_heredoc_indent` in MRI's parse.y... it even does tab conversion (WHY? No clue).

File: lib/ruby_lexer.rb (hunk 1 above)
Comment (jjoos) [inline lib/ruby_lexer.rb]: `indented content` I tried to follow the fixed width constant names, but certainly not ideal.

File: lib/ruby_lexer.rb (hunk 2 above)
Comment (jjoos) [inline lib/ruby_lexer.rb]: Ok will revert.

**Comment by jjoos** [comment]
Lemme try to make it equivalent to mri. I missed it because I looked at ruby/spec ( https://github.com/ruby/spec/blob/29b472cf57946ce271533fc6e03716908a313aaf/language/heredoc_spec.rb#L56 ) which doesn't test the tab conversion.

File: lib/ruby_lexer.rb (hunk 1 above)
Comment (jjoos) [inline lib/ruby_lexer.rb]: Btw got it from here: https://ruby-doc.org/core-2.3.0/doc/syntax/literals_rdoc.html#label-Here+Documents (search for "indented content" in that section).

**Comment by jjoos** [comment]
Updated: https://github.com/seattlerb/ruby_parser/pull/243/commits/5783a1591c27b34efa3527bbfade95e7887ccb44

**Comment by jjoos** [comment]
rebased commit with changes: 59d94904147b812736f16df9932db917208ae8ff

**Comment by guigs** [comment]
How can I test this in my project? 

I've tried adding to `Gemfile`:

````ruby
gem 'ruby_parser', '3.8.4', git: 'https://github.com/jjoos/ruby_parser', branch: 'squiggly-heredoc-support'
````

But then I get `LoadError: cannot load such file -- ruby_lexer.rex` in `ruby_lexer.rb:1305`.

**Comment by jjoos** [comment]
@guigs you need to clone the repo, make sure you have all dependencies installed (check https://github.com/seattlerb/ruby_parser/pull/244/files for details), run `rake` and point in your gemfile to the repo with `path: to/local/repo`

**Comment by zenspider** [comment]
I'm not thrilled with how complex this is, or the naming... but I'm adding as-is.

Thanks for the PR. Sorry for dragging my feet. I've been nose down for a while.

**Comment by simonc** [comment]
Hello :) Sorry to bother you but do you plan on releasing this soon? It breaks flog and thus rubycritic 😢 
Thanks for all the work you do ❤️

**Pull request closed without merging**